code/game/machinery/dish_drive.dm

```diff
@@ -0,0 +1,143 @@
+	var/list/dish_drive_contents
+
+/obj/machinery/dna_scannernew/Initialize(mapload)
+	. = ..()
+	component_parts = list()
+	component_parts += new /obj/item/stock_parts/manipulator(null)
+	component_parts += new /obj/item/stock_parts/matter_bin(null)
+	component_parts += new /obj/item/stack/sheet/glass(null)
+	component_parts += new /obj/item/stack/cable_coil(null, 1)
+	RefreshParts()
+
+/obj/machinery/dish_drive/examine(mob/user)
+	. = ..()
+	if(user.Adjacent(src))
+		. += "<span class='notice'>Alt-click it to beam its waste storage contents to any nearby disposal bins.</span>"
+
+/obj/machinery/dish_drive/attack_hand(mob/living/user)
+	if(!LAZYLEN(dish_drive_contents))
+		to_chat(user, "<span class='warning'>There's nothing in [src]!</span>")
+		return
+	var/obj/item/I = LAZYACCESS(dish_drive_contents, LAZYLEN(dish_drive_contents)) //the most recently-added item
+	LAZYREMOVE(dish_drive_contents, I)
+	user.put_in_hands(I)
+	to_chat(user, "<span class='notice'>You take out [I] from [src].</span>")
+	playsound(src, 'sound/items/pshoom.ogg', 50, TRUE)
+	flick("synthesizer_beam", src)
+
+/obj/machinery/dish_drive/attackby(obj/item/I, mob/living/user, params)
+	if(is_type_in_list(I, collectable_items) && user.a_intent != INTENT_HARM)
+		if(!user.drop_item())
+			return
+		LAZYADD(dish_drive_contents, I)
+		to_chat(user, "<span class='notice'>You put [I] in [src], and it's beamed into energy!</span>")
+		playsound(src, 'sound/items/pshoom.ogg', 50, TRUE)
+		flick("synthesizer_beam", src)
+		return
+	else if(default_deconstruction_screwdriver(user, "[initial(icon_state)]-o", initial(icon_state), I))
+		return
+	else if(default_unfasten_wrench(user, I))
+		return
+	else if(default_deconstruction_crowbar(I, FALSE))
+		return
+	..()
+
+/obj/machinery/dish_drive/RefreshParts()
+	idle_power_usage = initial(idle_power_usage)
+	active_power_usage = initial(active_power_usage)
+	use_power = initial(use_power)
+	var/total_rating = 0
+	for(var/obj/item/stock_parts/S in component_parts)
+		total_rating += S.rating
+	if(total_rating >= 9)
+		active_power_usage = 0
+		use_power = NO_POWER_USE
+	else
+		idle_power_usage = max(0, idle_power_usage - total_rating)
+		active_power_usage = max(0, active_power_usage - total_rating)
+	var/obj/item/circuitboard/dish_drive/board = locate() in component_parts
+	if(board)
+		suction_enabled = board.suction
+		transmit_enabled = board.transmit
+
+/obj/machinery/dish_drive/process()
+	if(time_since_dishes <= world.time && transmit_enabled)
+		do_the_dishes()
+	if(!suction_enabled)
+		return
+	for(var/obj/item/I in view(4, src))
+		if(is_type_in_list(I, collectable_items) && I.loc != src && (!I.reagents || !I.reagents.total_volume))
+			if(I.Adjacent(src))
+				LAZYADD(dish_drive_contents, I)
+				visible_message("<span class='notice'>[src] beams up [I]!</span>")
+				I.forceMove(src)
+				playsound(src, 'sound/items/pshoom.ogg', 50, TRUE)
+				flick("synthesizer_beam", src)
+			else
+				step_towards(I, src)
```

Comment:
Since we're dealing with `view()`, how does this behave when the dish drive tries to pull something from behind a window/windoor/etc?

Reply:
If the object is behind a window. It will continue to be pulled, then stop at the window and stay there until the obstacle is removed.